// detect-evm-proxy/build/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("./constants");
const helpers_1 = require("./helpers");
const types_1 = require("./types");
async function detectProxyTarget(proxyAddress, jsonRpcRequest, blockTag = "latest") {
    try {
        return await Promise.any([
            // EIP-1167 Minimal Proxy Contract
            jsonRpcRequest({
                method: "eth_getCode",
                params: [proxyAddress, blockTag],
            })
                .then(helpers_1.parse1167Bytecode)
                .then((val) => {
                const contractAddress = (0, helpers_1.readAddress)(val);
                return {
                    contractAddress,
                    kind: types_1.PROXY_KIND.EIP_1167_MINIMAL_PROXY_CONTRACT,
                };
            }),
            // EIP-1967 direct proxy
            jsonRpcRequest({
                method: "eth_getStorageAt",
                params: [proxyAddress, constants_1.EIP_1967_LOGIC_SLOT, blockTag],
            }).then((val) => {
                const contractAddress = (0, helpers_1.readAddress)(val);
                return {
                    contractAddress,
                    kind: types_1.PROXY_KIND.EIP_1967_TRANSPARENT_PROXY_PATTERN,
                };
            }),
            // EIP-1967 beacon proxy
            jsonRpcRequest({
                method: "eth_getStorageAt",
                params: [proxyAddress, constants_1.EIP_1967_BEACON_SLOT, blockTag],
            })
                .then(helpers_1.readAddress)
                .then((beaconAddress) => jsonRpcRequest({
                method: "eth_call",
                params: [
                    {
                        to: beaconAddress,
                        data: constants_1.EIP_1167_BEACON_METHODS[0],
                    },
                    blockTag,
                ],
            }).catch(() => jsonRpcRequest({
                method: "eth_call",
                params: [
                    {
                        to: beaconAddress,
                        data: constants_1.EIP_1167_BEACON_METHODS[1],
                    },
                    blockTag,
                ],
            })))
                .then((val) => {
                const contractAddress = (0, helpers_1.readAddress)(val);
                return {
                    contractAddress,
                    kind: types_1.PROXY_KIND.EIP_1967_TRANSPARENT_PROXY_PATTERN,
                };
            }),
            // EIP-897 DelegateProxy pattern
            jsonRpcRequest({
                method: "eth_call",
                params: [
                    {
                        to: proxyAddress,
                        data: constants_1.EIP_897_INTERFACE[0],
                    },
                    blockTag,
                ],
            }).then((val) => {
                const contractAddress = (0, helpers_1.readAddress)(val);
                return {
                    contractAddress,
                    kind: types_1.PROXY_KIND.EIP_897_DELEGATE_PROXY_PATTERN,
                };
            }),
            // EIP-1822 Universal Upgradeable Proxy Standard
            jsonRpcRequest({
                method: "eth_getStorageAt",
                params: [proxyAddress, constants_1.EIP_1822_LOGIC_SLOT, blockTag],
            }).then((val) => {
                const contractAddress = (0, helpers_1.readAddress)(val);
                return {
                    contractAddress,
                    kind: types_1.PROXY_KIND.EIP_1822_UNIVERSAL_UPGRADEABLE_PROXY_STANDARD,
                };
            }),
            // OpenZeppelin proxy pattern
            jsonRpcRequest({
                method: "eth_getStorageAt",
                params: [proxyAddress, constants_1.OPEN_ZEPPELIN_IMPLEMENTATION_SLOT, blockTag],
            }).then((val) => {
                const contractAddress = (0, helpers_1.readAddress)(val);
                return {
                    contractAddress,
                    kind: types_1.PROXY_KIND.OPEN_ZEPPELIN_PROXY_PATTERN,
                };
            }),
            // GnosisSafeProxy contract
            jsonRpcRequest({
                method: "eth_call",
                params: [
                    {
                        to: proxyAddress,
                        data: constants_1.GNOSIS_SAFE_PROXY_INTERFACE[0],
                    },
                    blockTag,
                ],
            }).then((val) => {
                const contractAddress = (0, helpers_1.readAddress)(val);
                return {
                    contractAddress,
                    kind: types_1.PROXY_KIND.GNOSIS_SAFE_PROXY_CONTRACT,
                };
            }),
        ]);
    }
    catch (_a) {
        return null;
    }
}
exports.default = detectProxyTarget;
